File: benchmarks/cpp/bench_baselines.cpp
#include <algorithm>
#include <cinttypes>
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <time.h>
#include <vector>

static volatile double g_sink = 0.0;

enum class Scenario {
    SumDense,
    ListUpdate,
    DotProduct,
    BranchMix,
};

static uint64_t now_ns() {
    timespec ts{};
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return static_cast<uint64_t>(ts.tv_sec) * 1000000000ull +
           static_cast<uint64_t>(ts.tv_nsec);
}

static size_t parse_size(const char *value, size_t fallback) {
    if (value == nullptr) {
        return fallback;
    }
    const auto parsed = static_cast<size_t>(std::strtoull(value, nullptr, 10));
    return parsed == 0 ? fallback : parsed;
}

static bool parse_scenario(const char *name, Scenario &scenario) {
    if (std::strcmp(name, "sum_dense") == 0) {
        scenario = Scenario::SumDense;
        return true;
    }
    if (std::strcmp(name, "list_update") == 0) {
        scenario = Scenario::ListUpdate;
        return true;
    }
    if (std::strcmp(name, "dot_product") == 0) {
        scenario = Scenario::DotProduct;
        return true;
    }
    if (std::strcmp(name, "branch_mix") == 0) {
        scenario = Scenario::BranchMix;
        return true;
    }
    return false;
}

static void fill_input(double *arr, size_t n) {
    for (size_t i = 0; i < n; ++i) {
        arr[i] = static_cast<double>(i);
    }
}

static double run_kernel(Scenario scenario, double *arr, size_t n, size_t reps) {
    double total = 0.0;
    switch (scenario) {
    case Scenario::SumDense:
        for (size_t r = 0; r < reps; ++r) {
            double sum = 0.0;
            for (size_t i = 0; i < n; ++i) {
                sum += 0.0;
                sum += 0.0;
                sum += arr[i];
            }
            total += sum;
        }
        break;
    case Scenario::ListUpdate:
        for (size_t r = 0; r < reps; ++r) {
            double sum = 0.0;
            for (size_t i = 0; i < n; ++i) {
                const double value = arr[i];
                sum += value;
                arr[i] = value + 1.0;
            }
            total += sum;
        }
        break;
    case Scenario::DotProduct:
        for (size_t r = 0; r < reps; ++r) {
            double sum = 0.0;
            for (size_t i = 0; i < n; ++i) {
                const double value = arr[i];
                sum += value * value;
            }
            total += sum;
        }
        break;
    case Scenario::BranchMix: {
        double sum = 0.0;
        int64_t state = 0;
        for (size_t r = 0; r < reps; ++r) {
            for (size_t i = 0; i < n; ++i) {
                const double value = arr[i];
                state += 17;
                if (state >= 97) {
                    state -= 97;
                }
                if (state < 48) {
                    sum += value;
                } else {
                    sum -= value;
                }
            }
        }
        return sum;
    }
    }
    return total;
}

static double cv_pct(const std::vector<uint64_t> &samples) {
    if (samples.size() < 2) {
        return 0.0;
    }
    long double mean = 0.0;
    for (const uint64_t sample : samples) {
        mean += static_cast<long double>(sample);
    }
    mean /= static_cast<long double>(samples.size());
    if (mean == 0.0) {
        return 0.0;
    }
    long double variance = 0.0;
    for (const uint64_t sample : samples) {
        const long double delta = static_cast<long double>(sample) - mean;
        variance += delta * delta;
    }
    variance /= static_cast<long double>(samples.size());
    return std::sqrt(static_cast<double>(variance)) * 100.0 /
           static_cast<double>(mean);
}

int main(int argc, char **argv) {
    if (argc < 2) {
        std::fprintf(stderr, "usage: %s <sum_dense|list_update|dot_product|branch_mix> [n] [iters] [warmup_ms] [reps]\n", argv[0]);
        return 2;
    }

    Scenario scenario = Scenario::SumDense;
    if (!parse_scenario(argv[1], scenario)) {
        std::fprintf(stderr, "unknown scenario: %s\n", argv[1]);
        return 2;
    }

    const size_t n = parse_size(argc > 2 ? argv[2] : nullptr, 100000);
    const size_t iters = parse_size(argc > 3 ? argv[3] : nullptr, 200);
    const size_t warmup_ms = parse_size(argc > 4 ? argv[4] : nullptr, 100);
    const size_t reps = parse_size(argc > 5 ? argv[5] : nullptr, 50);

    std::vector<uint64_t> samples(iters);

    const uint64_t warmup_end = now_ns() + warmup_ms * 1000000ull;
    while (now_ns() < warmup_end) {
        std::vector<double> arr(n);
        fill_input(arr.data(), n);
        g_sink = run_kernel(scenario, arr.data(), n, reps);
    }

    double checksum = 0.0;
    for (size_t iteration = 0; iteration < iters; ++iteration) {
        const uint64_t start = now_ns();
        {
            std::vector<double> arr(n);
            fill_input(arr.data(), n);
            checksum = run_kernel(scenario, arr.data(), n, reps);
            g_sink = checksum;
        }
        samples[iteration] = now_ns() - start;
    }

    const double sample_cv_pct = cv_pct(samples);
    std::sort(samples.begin(), samples.end());
    const uint64_t median = samples[iters / 2];
    const uint64_t p95 = samples[(iters * 95) / 100];
    const double ops = median == 0 ? 0.0 : 1e9 / static_cast<double>(median);

    std::printf(
        "[CPP BENCH] %s median=%" PRIu64
        " ns/op (%.0f ops/sec), p95=%" PRIu64
        " ns/op cv_pct=%.4f | iters=%zu warmup=%zums reps=%zu checksum=%.17g\n",
        argv[1],
        median,
        ops,
        p95,
        sample_cv_pct,
        iters,
        warmup_ms,
        reps,
        checksum);
    return 0;
}

File: benchmarks/zig/bench_baselines.zig
const std = @import("std");

var result_sink: f64 = 0.0;

fn nowNs(io: std.Io) i96 {
    return std.Io.Clock.awake.now(io).nanoseconds;
}

fn runScenario(name: []const u8, a: []f64, reps: usize) !f64 {
    for (a, 0..) |*value, i| {
        value.* = @floatFromInt(i);
    }

    var total: f64 = 0.0;
    if (std.mem.eql(u8, name, "sum_dense")) {
        for (0..reps) |_| {
            var sum: f64 = 0.0;
            for (a) |value| {
                sum += value;
            }
            total += sum;
        }
    } else if (std.mem.eql(u8, name, "list_update")) {
        for (0..reps) |_| {
            var sum: f64 = 0.0;
            for (a) |*value| {
                const current = value.*;
                sum += current;
                value.* = current + 1.0;
            }
            total += sum;
        }
    } else if (std.mem.eql(u8, name, "dot_product")) {
        for (0..reps) |_| {
            var sum: f64 = 0.0;
            for (a) |value| {
                sum += value * value;
            }
            total += sum;
        }
    } else if (std.mem.eql(u8, name, "branch_mix")) {
        var sum: f64 = 0.0;
        var state: i64 = 0;
        for (0..reps) |_| {
            for (a) |value| {
                state += 17;
                if (state >= 97) {
                    state -= 97;
                }
                if (state < 48) {
                    sum += value;
                } else {
                    sum -= value;
                }
            }
        }
        return sum;
    } else {
        return error.UnknownBenchmark;
    }
    return total;
}

fn parseArg(arg: ?[]const u8, fallback: usize) !usize {
    const value = arg orelse return fallback;
    return std.fmt.parseInt(usize, value, 10);
}

fn coefficientOfVariationPct(samples: []const u64) f64 {
    if (samples.len < 2) return 0.0;
    var mean: f64 = 0.0;
    for (samples) |sample| {
        mean += @floatFromInt(sample);
    }
    mean /= @floatFromInt(samples.len);
    if (mean == 0.0) return 0.0;

    var variance: f64 = 0.0;
    for (samples) |sample| {
        const delta: f64 = @as(f64, @floatFromInt(sample)) - mean;
        variance += delta * delta;
    }
    variance /= @floatFromInt(samples.len);
    return @sqrt(variance) * 100.0 / mean;
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    var args = try std.process.Args.Iterator.initAllocator(init.minimal.args, allocator);
    defer args.deinit();

    _ = args.next();
    const name = args.next() orelse return error.MissingBenchmark;
    const n = try parseArg(args.next(), 100_000);
    const iters = try parseArg(args.next(), 50);
    const warmup_ms = try parseArg(args.next(), 100);
    const reps = try parseArg(args.next(), 50);
    if (iters == 0) {
        return error.ZeroIterations;
    }

    const samples = try allocator.alloc(u64, iters);
    defer allocator.free(samples);

    const warmup_ns: i96 = @intCast(warmup_ms * std.time.ns_per_ms);
    const warmup_end = nowNs(init.io) + warmup_ns;
    while (nowNs(init.io) < warmup_end) {
        const a = try allocator.alloc(f64, n);
        const checksum = runScenario(name, a, reps) catch |err| {
            allocator.free(a);
            return err;
        };
        const sink: *volatile f64 = &result_sink;
        sink.* = checksum;
        allocator.free(a);
    }

    var checksum: f64 = 0.0;
    for (samples) |*sample| {
        const start = nowNs(init.io);
        const a = try allocator.alloc(f64, n);
        checksum = runScenario(name, a, reps) catch |err| {
            allocator.free(a);
            return err;
        };
        const sink: *volatile f64 = &result_sink;
        sink.* = checksum;
        allocator.free(a);
        sample.* = @intCast(nowNs(init.io) - start);
    }

    const cv_pct = coefficientOfVariationPct(samples);
    var i: usize = 0;
    while (i < samples.len) : (i += 1) {
        var j = i + 1;
        while (j < samples.len) : (j += 1) {
            if (samples[j] < samples[i]) {
                const tmp = samples[i];
                samples[i] = samples[j];
                samples[j] = tmp;
            }
        }
    }

    const median = samples[samples.len / 2];
    const p95 = samples[(samples.len * 95) / 100];
    std.debug.print(
        "[ZIG BENCH] {s} median={d} ns/op, p95={d} ns/op cv_pct={d:.4} checksum={d:.17}\n",
        .{ name, median, p95, cv_pct, checksum },
    );
}
